File: Chess.php
<?php
	session_start();
?>
<title>Chess</title>
<tt>
<?php
	class Square
	{
		private $color;
		private $type;
		private $moved;
		function __construct($color, $type)
		{
			$this->color=$color;
			$this->type=$type;
			$this->moved=$color==' ';
		}
		function color()
		{
			return $this->color;
		}
		function type()
		{
			return $this->type;
		}
		function moved()
		{
			return $this->moved;
		}
		function move()
		{
			$this->moved=true;
		}
		function promote()
		{
			$this->color="Qn";
		}
		function equals($square)
		{
			return $this->color==$square->color&&$this->type==$square->type;
		}
		function toString()
		{
			return $this->color.$this->type;
		}
	}
	class Chess
	{
		private $boards;
		private $moves;
		private $epr;
		private $epc;
		private $board;
		private $wkr;
		private $wkc;
		private $bkr;
		private $bkc;
		private $turn;
		private $cap;
		function __construct()
		{
			$pieces=["Rk", "Kt", "Bp", "Kg", "Qn", "Bp", "Kt", "Rk"];
			$this->boards=[];
			$this->moves=0;
			$this->epr=8;
			$this->epc=8;
			$this->setVariables([], 0, 3, 7, 3, "W", "B");
			for($r=0; $r<8; $r++)
			{
				$this->board[$r]=[];
				for($c=0; $c<8; $c++)
					$this->board[$r][$c]=new Square(' ', "  ");
			}
			for($c=0; $c<8; $c++)
			{
		  		$this->board[0][$c]=new Square("W", $pieces[$c]);
		  		$this->board[1][$c]=new Square("W", "Pn");
		  		$this->board[6][$c]=new Square("B", "Pn");
		  		$this->board[7][$c]=new Square("B", $pieces[$c]);
		  	}
		}
		function boardCopy()
		{
			$board=[];
			for($r=0; $r<8; $r++)
			{
				$board[$r]=[];
				for($c=0; $c<8; $c++)
					$board[$r][$c]=new Square($this->board[$r][$c]->color(), $this->board[$r][$c]->type());
			}
			return $board;
		}
		function setVariables($board, $wkr, $wkc, $bkr, $bkc, $turn, $cap)
		{
			$this->board=$board;
			$this->wkr=$wkr;
			$this->wkc=$wkc;
			$this->bkr=$bkr;
			$this->bkc=$bkc;
			$this->turn=$turn;
			$this->cap=$cap;
		}
		function changeTurns()
		{
			$t=$this->turn;
			$this->turn=$this->cap;
			$this->cap=$t;
		}
		function canPawnMove($r1, $c1, $r2, $c2)
		{
			$r=$r1-$r2;
			$a=abs($r);
			$c=abs($c1-$c2);
		  	if($r==0||$a>2||$r>0&&$this->turn=='W'||$r<0&&$this->turn=='B')
		  		return false;
		  	if($c==1)
		  		return $a==1&&($this->board[$r2][$c2]->color()==$this->cap||$r1==$this->epr&&c2==$this->epc);
		  	if($c==0&&$this->board[$r2][$c2]->color()==' ')
		  		return $a==1||$this->board[$r2+$r/2][$c2]->color()==' '&&!$this->board[$r1][$c1]->moved();
		  	return false;
		}
		function canKnightMove($r1, $c1, $r2, $c2)
		{
			$r=abs($r1+$r2);
			$c=abs($c1-$c2);
			return $r<3&&$c<3&&$r+$c==3;
		}
		function canRookMove($r1, $c1, $r2, $c2)
		{
			if($r1!=$r2&&$c1!=$c2)
				return false;
			if($r1>$r2)
			{
				$t=$r1;
				$r1=$r2;
				$r2=$t;
			}
			else if($c1>$c2)
			{
				$t=$c1;
				$c1=$c2;
				$c2=$t;
			}
			if($r1!=$r2)
				for($r=$r1+1; $r<$r2; $r++)
				{
					if($this->board[$r][$c1]->color()!=' ')
						return false;
				}
			else
				for($c=$c1+1; $c<$c2; $c++)
					if($this->board[$r1][$c]->color()!=' ')
						return false;
			return true;
		}
		function canBishopMove($r1, $c1, $r2, $c2)
		{
			if($r1>$r2)
			{
				$t=$r1;
				$r1=$r2;
				$r2=$t;
				$t=$c1;
				$c1=$c2;
				$c2=$t;
			}
			$r=$r2-$r1;
			$c=$c1-$c2;
			if($r!=abs($c))
				return false;
			if($c<0)
				for($i=1; $i<$r; $i++)
				{
					if($this->board[$r1+$i][$c1+$i]->color()!=' ')
						return false;
				}
			else
				for($i=1; $i<$r; $i++)
					if($this->board[$r1+$i][$c1-$i]->color()!=' ')
						return false;
			return true;
		}
		function canKingMove($r1, $c1, $r2, $c2)
		{
			$r=abs($r1-$r2);
			$c=$c1-$c2;
			$a=abs($c);
			if($a<2)
				return $r<2;
			if($r>0||$this->board[$r1][$c1]->moved()||$a!=2)
				return false;
			if($c==2)
				return $this->board[$r2][$c1-1]->color()==' '&&$this->board[$r2][$c2]->color()==' '&&!$this->board[$r2][0]->moved();
			return $this->board[$r2][$c1+1]->color()==' '&&$this->board[$r2][$c2]->color()==' '&&$this->board[$r2][$c1+2]->color()==' '&&!$this->board[$r2][7]->moved();
		}
		function move($r1, $c1, $r2, $c2)
		{
			array_push($this->boards, $this->boardCopy());
			if($this->board[$r1][$c1]->type()=="Kg")
			{
				$this->epr=8;
				if($this->turn=='W')
				{
					$this->wkr=$r2;
					$this->wkc=$c2;
				}
				else
				{
					$this->bkr=$r2;
					$this->bkc=$c2;
				}
				if(abs($c1-$c2)==2)
					if($c2==1)
					{
						$this->board[$r2][2]=new Square($this->board[$r2][0]->color(), $this->board[$r2][0]->type());
						$this->board[$r2][0]=new Square(' ', "  ");
					}
					else
					{
						$this->board[$r2][4]=new Square($this->board[$r2][7]->color(), $this->board[$r2][7]->type());
						$this->board[$r2][7]=new Square(' ', "  ");
					}
			}
			else if($this->board[$r1][$c1]->type()=="Pn")
			{
				$this->moves=-1;
				if(abs($r1-$r2)==2)
				{
					$this->epr=$r2;
					$this->epc=$c2;
				}
				else
				{
					if($r1==$this->epr&&$c2==$this->epc)
						$this->board[$r1][$c2]=new Square(' ', "  ");
					else if($r2==0||$r2==7)
						$this->board[$r1][$c1]->promote();
					$this->epr=8;
				}
			}
			else
				$this->epr=8;
			if($this->board[$r2][$c2]->color()==" ")
				$this->moves++;
			else
				$this->moves=0;
			$this->board[$r2][$c2]=new Square($this->board[$r1][$c1]->color(), $this->board[$r1][$c1]->type());
			$this->board[$r2][$c2]->move();
			$this->board[$r1][$c1]=new Square(' ', "  ");
			$this->changeTurns();
		}
		function canMove($r1, $c1, $r2, $c2, $real)
		{
			if($this->board[$r1][$c1]->color()!=$this->turn||$this->board[$r2][$c2]->color()==$this->turn)
				return false;
			if($real)
			{
				$chess=new Chess();
				$chess->setVariables($this->boardCopy(), $this->wkr, $this->wkc, $this->bkr, $this->bkc, $this->turn, $this->cap);
				if($this->board[$r1][$c1]->type()=="Kg"&&abs($c1-$c2)==2)
				{
			  		$chess->changeTurns();
			  		if($chess->anyCanMoveTo($r1, $c1, false))
			  			return false;
			  		$chess->changeTurns();
			  		$c=$c1+($c2-$c1)/2;
			  		$chess->move($r1, $c1, $r2, $c);
			  		if($chess->anyCanMoveTo($r2, $c, false))
			  			return false;
			  		$chess->move($r2, $c, $r1, $c1);
			  	}
			  	$chess->move($r1, $c1, $r2, $c2);
			  	if($this->turn=='W')
			  	{
			  		if($chess->anyCanMoveTo($chess->wkr, $chess->wkc, false))
			  			return false;
			  	}
			  	else if($chess->anyCanMoveTo($chess->bkr, $chess->bkc, false))
			  		return false;
			}
			if($this->board[$r1][$c1]->type()=="Pn")
				return $this->canPawnMove($r1, $c1, $r2, $c2);
			else if($this->board[$r1][$c1]->type()=="Kt")
				return $this->canKnightMove($r1, $c1, $r2, $c2);
			else if($this->board[$r1][$c1]->type()=="Rk")
				return $this->canRookMove($r1, $c1, $r2, $c2);
			else if($this->board[$r1][$c1]->type()=="Bp")
				return $this->canBishopMove($r1, $c1, $r2, $c2);
			else if($this->board[$r1][$c1]->type()=="Qn")
				return $this->canRookMove($r1, $c1, $r2, $c2)||$this->canBishopMove($r1, $c1, $r2, $c2);
			return $this->canKingMove($r1, $c1, $r2, $c2);
		}
		function anyCanMoveTo($r2, $c2, $real)
		{
			for($r=0; $r<8; $r++)
				for($c=0; $c<8; $c++)
					if($this->canMove($r, $c, $r2, $c2, $real))
						return true;
			return false;
		}
		function anyCanMove()
		{
			for($r=0; $r<8; $r++)
				for($c=0; $c<8; $c++)
					if($this->anyCanMoveTo($r, $c, true))
						return true;
			return false;
		}
		function equals($board)
		{
			for($r=0; $r<8; $r++)
				for($c=0; $c<8; $c++)
					if(!$this->board[$r][$c]->equals($board[$r][$c]))
						return false;
			return true;
		}
		function state()
		{
			if($this->moves==100)
				return "Draw by 50-move rule";
			$s="";
			$n=0;
			$w=0;
			$b=0;
			$this->changeTurns();
			if($this->turn=='W'&&$this->anyCanMoveTo($this->bkr, $this->bkc, false)||$this->turn=='B'&&$this->anyCanMoveTo($this->wkr, $this->wkc, false))
				$s="Check";
			$this->changeTurns();
			if(!$this->anyCanMove())
				return $s=="Check"?"Checkmate":"Stalemate";
			foreach($this->boards as $board)
				if($this->equals($board))
					$n++;
			if($n==2)
				return "Draw by 3-fold repetition";
			for($r=0; $r<8; $r++)
				for($c=0; $c<8; $c++)
					if($this->board[$r][$c]->type()=="Rk"||$this->board[$r][$c]->type()=="Qn"||$this->board[$r][$c]->type()=="Pn")
						return $s;
					else if($this->board[$r][$c]->color()=='W')
						$w++;
					else if($this->board[$r][$c]->color()=='B')
						$b++;
			return $w<2&&$b<2?"Draw by insufficient material":$s;
		}
		function randomMove()
		{
			$moves=[];
			for($r1=0; $r1<8; $r1++)
				for($c1=0; $c1<8; $c1++)
					for($r2=0; $r2<8; $r2++)
						for($c2=0; $c2<8; $c2++)
							if($this->canMove($r1, $c1, $r2, $c2, true))
								array_push($moves, [$r1, $c1, $r2, $c2]);
			$move=$moves[rand(0, count($moves)-1)];
			$this->move($move[0], $move[1], $move[2], $move[3]);
		}
		function toString()
		{
			$s=" H  G  F  E  D  C  B  A<br/>";
			for($r=0; $r<8;)
			{
				for($c=0; $c<8; $c++)
					$s=$s.$this->board[$r][$c]->toString();
				$s=$s." ".++$r."<br/>";
			}
			return str_replace(" ", "&nbsp;", $s);
		}
	}
	$state=isset($_SESSION["state"])?$_SESSION["state"]:"";
	$move=isset($_POST['move'])?$_POST['move']:"MASON";
	if(isset($_POST['reset']))
	{
		$_SESSION["chess"]=new Chess();
		$_SESSION["state"]="";
		echo($_SESSION["chess"]->toString());
	}
	else if(strlen($move)!=5)
		echo("That is not a valid move!");
	else if(isset($_POST['submit'])&&!(str_starts_with($state, "Draw")||str_ends_with($state, "mate")))
	{
		$r1=$move[1]-1;
		$c1=104-ord($move[0]);
		$r2=$move[4]-1;
		$c2=104-ord($move[3]);
		$chess=$_SESSION["chess"];
		if($chess->canMove($r1, $c1, $r2, $c2, true))
		{
			$chess->move($r1, $c1, $r2, $c2);
			$state=$chess->state();
			if(!(str_starts_with($state, "Draw")||str_ends_with($state, "mate")))
			{
				$chess->randomMove();
				$state=$chess->state();
			}
		}
		else
			$state="That is not a legal move!";
		$_SESSION["state"]=$state;
		echo($chess->toString());
		echo($state);
	}
	else
		echo "You are not in a game!";
?>
</tt>
<form method="POST">
	<input type='submit' name="reset" value="Reset"/>
</form>
<form method="POST">
	<input type='text' id="move" name="move"/><br/>
	<input type='submit' name='submit' value='Move'/>
</form>
<script type="text/javascript">
	document.getElementById('move').focus();
</script>

File: Chess.rb
class Square
	def initialize(color, type)
		@color=color
		@type=type
		@moved=color==' '
	end
	def equals(square)
		return @color==square.color()&&@type==square.type()
	end
	def promote()
		@type="Qn"
	end
	def move()
		@moved=true
	end
	def color()
		return @color
	end
	def type()
		return @type
	end
	def moved()
		return @moved
	end
	def toString()
		return @color+@type
	end
end
class Chess
	def initialize()
		pieces=["Rk", "Kt", "Bp", "Kg", "Qn", "Bp", "Kt", "Rk"]
		setVariables([], 0, 3, 7, 3, 'W', 'B')
		for r in 0..7
			@board[r]=[]
			for c in 0..7
				@board[r][c]=Square.new(' ', "  ")
			end
		end
		for c in 0..7
			@board[0][c]=Square.new('W', pieces[c])
			@board[1][c]=Square.new('W', "Pn")
			@board[6][c]=Square.new('B', "Pn")
			@board[7][c]=Square.new('B', pieces[c])
		end
		@moves=0
	end
	def boardCopy()
		board=[]
		for r in 0..7
			board[r]=[]
			for c in 0..7
				board[r][c]=Square.new(@board[r][c].color(), @board[r][c].type())
			end
		end
		return board
	end
	def setVariables(board, wkr, wkc, bkr, bkc, turn, cap)
		@board=board
		@wkr=wkr
		@wkc=wkc
		@bkr=bkr
		@bkc=bkc
		@turn=turn
		@cap=cap
		@boards=[]
		@epr=8
	end
	def changeTurns()
		t=@turn
		@turn=@cap
		@cap=t
	end
	def canPawnMove(r1, c1, r2, c2)
		r=r1-r2
		a=r.abs()
		c=(c1-c2).abs()
		if(r==0||a>2||r>0&&@turn=='W'||r<0&&@turn=='B')
			return false
		end
		if(c==1)
			return a==1&&(@board[r2][c2].color()==@cap||r1==@epr&&c2==@epc)
		end
		if(c==0&&@board[r2][c2].color()==' ')
			return a==1||(@board[r2+r/2][c2].color()==' '&&(not @board[r1][c1].moved()))
		end
		return false
	end
	def canKnightMove(r1, c1, r2, c2)
		r=(r1-r2).abs()
		c=(c1-c2).abs()
		return r<3&&c<3&&r+c==3
	end
	def canRookMove(r1, c1, r2, c2)
		if(r1!=r2&&c1!=c2)
			return false
		end
		if(r1>r2)
			t=r1
			r1=r2
			r2=t
		elsif(c1>c2)
			t=c1
			c1=c2
			c2=t
		end
		if(r1!=r2)
			for r in r1+1...r2
				if(@board[r][c2].color()!=' ')
					return false
				end
			end
		else
			for c in c1+1...c2
				if(@board[r2][c].color()!=' ')
					return false
				end
			end
		end
		return true
	end
	def canBishopMove(r1, c1, r2, c2)
		if(r1>r2)
			t=r1
			r1=r2
			r2=t
			t=c1
			c1=c2
			c2=t
		end
		r=r2-r1
		c=c1-c2
		if(r!=c.abs())
			return false
		end
		if(c<0)
			for i in 1...r
				if(@board[r1+i][c1+i].color()!=' ')
					return false
				end
			end
		else
			for i in 1...r
				if(@board[r1+i][c1-i].color()!=' ')
					return false
				end
			end
		end
		return true
	end
	def canKingMove(r1, c1, r2, c2)
		r=(r1-r2).abs()
		c=c1-c2
		a=c.abs()
		if(a<2)
			return r<2
		end
		if(r>0||@board[r1][c1].moved()||a!=2)
			return false
		end
		if(c==2)
			return @board[r2][c1-1].color()==' '&&@board[r2][c2].color()==' '&&!@board[r2][0].moved()
		end
		return @board[r2][c1+1].color()==' '&&@board[r2][c1+2].color()==' '&&@board[r2][c2].color()==' '&&!@board[r2][7].moved()
	end
	def move(r1, c1, r2, c2)
		@boards.push(boardCopy())
		if(@board[r1][c1].type()=="Kg")
			@epr=8
			if(@turn=='W')
				@wkr=r2
				@wkc=c2
			else
				@bkr=r2
				@bkc=c2
			end
			if((c1-c2).abs()==2)
				if(c2==1)
					@board[r2][2]=Square.new(@turn, "Rk")
					@board[r2][0]=Square.new(' ', "  ")
				else
					@board[r2][4]=Square.new(@turn, "Rk")
					@board[r2][7]=Square.new(' ', "  ")
				end
			end
		elsif(@board[r1][c1].type()=="Pn")
			@moves=-1
			if((r1-r2).abs()==2)
				@epr=r2
				@epc=c2
			else
				if(@epr==r1&&@epc==c2)
					@board[r1][c2]=Square.new(' ', "  ")
				elsif(r2==0||r2==7)
					@board[r2][c2].promote()
				end
				@epr=8
			end
		else
			@epr=8
		end
		if(@board[r2][c2].color()==' ')
			@moves+=1
		else
			@moves=0
		end
		@board[r2][c2]=Square.new(@turn, @board[r1][c1].type())
		@board[r2][c2].move()
		@board[r1][c1]=Square.new(' ', "  ")
		changeTurns()
	end
	def wkr()
		return @wkr
	end
	def wkc()
		return @wkc
	end
	def bkr()
		return @bkr
	end
	def bkc()
		return @bkc
	end
	def canMove(r1, c1, r2, c2, real)
		if(@board[r1][c1].color()!=@turn||@board[r2][c2].color()==@turn)
			return false
		end
		if(real)
			chess=Chess.new()
			chess.setVariables(boardCopy(), @wkr, @wkc, @bkr, @bkc, @turn, @cap)
			if(@board[r1][c1].type()=="Kg"&&(c1-c2).abs()==2)
				chess.changeTurns()
				if(chess.anyCanMoveTo(r1, c1, false))
					return false
				end
				chess.changeTurns()
				c=c1+(c2-c1)/2
				chess.move(r1, c1, r2, c)
				if(chess.anyCanMoveTo(r2, c, false))
					return false
				end
				chess.move(r2, c, r1, c1)
			end
			chess.move(r1, c1, r2, c2)
			if(@turn=='W')
				if(chess.anyCanMoveTo(chess.wkr(), chess.wkc(), false))
					return false
				end
			elsif(chess.anyCanMoveTo(chess.bkr(), chess.bkc(), false))
				return false
			end
		end
		if(@board[r1][c1].type=="Pn")
			return canPawnMove(r1, c1, r2, c2)
		elsif(@board[r1][c1].type=="Kt")
			return canKnightMove(r1, c1, r2, c2)
		elsif(@board[r1][c1].type=="Rk")
			return canRookMove(r1, c1, r2, c2)
		elsif(@board[r1][c1].type=="Bp")
			return canBishopMove(r1, c1, r2, c2)
		elsif(@board[r1][c1].type=="Qn")
			return canRookMove(r1, c1, r2, c2)||canBishopMove(r1, c1, r2, c2)
		end
		return canKingMove(r1, c1, r2, c2)
	end
	def anyCanMoveTo(r2, c2, real)
		for r in 0..7
			for c in 0..7
				if(canMove(r, c, r2, c2, real))
					return true
				end
			end
		end
		return false
	end
	def anyCanMove()
		for r in 0..7
			for c in 0..7
				if(anyCanMoveTo(r, c, true))
					return true
				end
			end
		end
		return false
	end
	def equals(board)
		for r in 0..7
			for c in 0..7
				if(!@board[r][c].equals(board[r][c]))
					return false
				end
			end
		end
		return true
	end
	def state()
		if(@moves==100)
			return "Draw by 50-move rule"
		end
	  	s=""
	  	n=0
	  	w=0
	  	b=0
	  	changeTurns()
	  	if(@turn=='W'&&anyCanMoveTo(@bkr, @bkc, false)||@turn=='B'&&anyCanMoveTo(@wkr, @wkc, false))
	  		s="Check"
	  	end
	  	changeTurns()
	  	if(!anyCanMove())
	  		if(s=="Check")
	  			return "Checkmate"
	  		end
	  		return "Stalemate"
	  	end
	  	for board in @boards
	  		if(equals(board))
	  			n+=1
	  		end
	  	end
	  	if(n==2)
	  		return "Draw by 3-fold repetition"
	  	end
	  	for r in 0..7
	  		for c in 0..7
	  			if(@board[r][c].type()=="Rk"||@board[r][c].type()=="Qn"||@board[r][c].type()=="Pn")
	  				return s
	  			elsif(@board[r][c].color()=='W')
	  				w+=1
	  			elsif(@board[r][c].color()=='B')
	  				b+=1
	  			end
	  		end
	  	end
	  	if(w<2&&b<2)
	  		return "Draw by insufficient material"
	  	end
	  	return s
	end
	def randomMove()
		moves=[]
		for r1 in 0..7
			for c1 in 0..7
				for r2 in 0..7
					for c2 in 0..7
						if(canMove(r1, c1, r2, c2, true))
							moves.push([r1, c1, r2, c2])
						end
					end
				end
			end
		end
		move=moves[(rand()*moves.length()).to_i()]
		move(move[0], move[1], move[2], move[3])
	end
	def toString()
		s=" h  g  f  e  d  c  b  a\n"
		for r in 0..7
			for c in 0..7
				s+=@board[r][c].toString()
			end
			s+=" "+(r+1).to_s()+"\n"
		end
		return s
	end
end
chess=Chess.new()
state=""
puts(chess.toString())
while(!(state.start_with?("Draw")||state.end_with?("mate")))
	b=true
	while(b)
		move=gets()
		r1=move[1].to_i()-1
		c1=104-move[0].ord()
		r2=move[4].to_i()-1
		c2=104-move[3].ord()
		if(r1<0||r1>7||c1<0||c1>7||r2<0||r2>7||c2<0||c2>7)
			puts("That is not a valid move!");
		elsif(chess.canMove(r1, c1, r2, c2, true))
			chess.move(r1, c1, r2, c2)
			state=chess.state()
			puts(chess.toString()+state)
			b=false
		else
			puts("That is not a legal move!")
		end
	end
	if(!(state.start_with?("Draw")||state.end_with?("mate")))
		chess.randomMove()
		state=chess.state()
		puts(chess.toString()+state)
	end
end